joinmarket/blockchaininterface.py

```diff
@@ -637,8 +637,9 @@ def sync_addresses(self, wallet):
                 if last_used_addr == '':
                     wallet.index[mix_depth][forchange] = 0
                 else:
-                    wallet.index[mix_depth][forchange] = \
-                        wallet.addr_cache[last_used_addr][2] + 1
+		    next_avail_idx = max([wallet.addr_cache[last_used_addr][2]+1,
+		                        wallet.index_cache[mix_depth][forchange]])
+		    wallet.index[mix_depth][forchange] = next_avail_idx
```

Comment:
The several tests (most commented out for build) in test_wallet_sync in test_wallets.py have indicated that this gives the "correct" result in spamming and non-spamming case: example: you do some transactions and use 3 addresses in a branch, then a spammer requests 4 more, this will return index 7 (i.e. the 8th, after the first 7 were used up).